File: LinkedList/SinglyCL.java
// SinglyCL  java:

import java.util.*;

class node
{
    public int data;
    public node next;
}

class Singly
{
    private node First;
    private node Last;
    private int iCount;

    public Singly()
    {
        System.out.println("Inside Singly Counstructor !!");
        First = null;
        Last = null;
        iCount = 0;
    }

    public void Display()
    {
        System.out.print("-> ");
        do
        {
            System.out.print("| "+First.data+" | -> ");
            First = First.next;
        }while(First != Last.next);

        System.out.println();
    }

    public int Count()
    {
        return iCount;
    }

    public void InsertFirst(int iNo)
    {
        node newn = null;

        newn = new node(); 
        newn.next = null;
        newn.data = iNo;

        if(First == null || Last == null )
        {
            Last = newn;
        }
        else
        {
            newn.next = First;
        }
        First = newn;
        Last.next = First;
        
        iCount++;
    }

    public void InsertLast(int iNo)
    {
        node newn = null;

        newn = new node();
        newn.next = null;
        newn.data = iNo;

        if(First == null || Last == null )
        {
            First = newn;
            Last = newn;
        }
        else
        {
            Last.next = newn;
            Last = newn;
        }
        Last.next = First;
        iCount++;
    }

    public void InsertatPos(int iPos, int iNo)
    {
        node temp = null;
        int i = 0;

        node newn = new node();
        newn.next = null;
        newn.data = iNo;

        if(iPos<1 || iPos>iCount+1)
        {
            System.out.println("Invalid Position !!");
            return;
        }
        
        if(iPos == iCount+1)
        {
            InsertLast(iNo);
        }
        else if(iPos == 1)
        {
            InsertFirst(iNo);
        }
        else 
        {
            temp = First;
            for(i=1; i<iPos-1; i++)
            {
                temp = temp.next;
            }
            newn.next = temp.next;
            temp.next = newn;

            iCount++;
        }
    }

    public void DeleteFirst()
    {
        if(First == null || Last == null )
        {
            System.out.println("LL is empty !!");
            return;
        }
        else if(First == Last)
        {
            First  = null;
            Last = null;
        }
        else
        {
            First = First.next;
            Last.next = First;
        }
        iCount--;
    }

    public void DeleteLast()
    {
        node temp = First;

        if(First == null || Last == null )
        {
            System.out.println("LL is empty !!");
            return;
        }
        else if(First == Last)
        {
            First  = null;
            Last = null;
        }
        else
        {
            while(temp.next != Last)
            {
                temp = temp.next;
            }
            Last = temp;
            Last.next = First;
        }
        iCount--;
    }

    public void DeleteAtPos(int iPos)
    {
        node temp = null;
        int i = 0;

        if(iPos<1 || iPos>iCount)
        {
            System.out.println("Invalid Position !!");
            return;
        }
        
        if(iPos == iCount)
        {
            DeleteLast();
        }
        else if(iPos == 1)
        {
            DeleteFirst();
        }
        else 
        {
            temp = First;
            for(i=1; i<iPos-1; i++)
            {
                temp = temp.next;
            }
            temp.next = temp.next.next;

            iCount--;
        }
    }

}

class SinglyCL 
{
    public static void main(String[] args) 
    { 
        try (Scanner Sobj = new Scanner(System.in)) 
        {
            Singly Aobj = new Singly();
            int iRet = 0;

            Aobj.InsertFirst(51);
            Aobj.InsertFirst(21);
            Aobj.InsertFirst(10);

            Aobj.InsertLast(101);
            Aobj.InsertLast(121);
            Aobj.InsertLast(151);

            Aobj.Display();
            iRet = Aobj.Count();
            System.out.println("Number of elements in LL are :"+iRet);

            System.out.println("Enter the Position where you want to insert the element :");
            int Index = Sobj.nextInt();

            Aobj.InsertatPos(Index, 15);

            Aobj.Display();
            iRet = Aobj.Count();
            System.out.println("Number of elements in LL are :"+iRet);

            Aobj.DeleteFirst();

            Aobj.Display();
            iRet = Aobj.Count();
            System.out.println("Number of elements in LL are :"+iRet);

            Aobj.DeleteLast();

            Aobj.Display();
            iRet = Aobj.Count();
            System.out.println("Number of elements in LL are :"+iRet);

            Aobj.DeleteAtPos(3);

            Aobj.Display();
            iRet = Aobj.Count();
            System.out.println("Number of elements in LL are :"+iRet);
        }

    }
}

File: LinkedList/SinglyCL.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

void Display(PNODE First, PNODE Last)
{
    printf("->");
    do
    {
        printf("| %d | ->",First->data);
        First = First->next;
    }while(First != Last->next);
    printf("\n");
}

int Count(PNODE First, PNODE Last)
{
    int iCount = 0;
    do
    {
        iCount++;
        First = First->next;
    }while(First != Last->next);

    return iCount;
}

void InsertFirst(PPNODE First, PPNODE Last, int No)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));
    newn->next = NULL;
    newn->data = No;

    if(*First == NULL || *Last == NULL)
    {
        *First = newn;
        *Last = newn;
    }
    else
    {
        newn->next = *First;
        *First = newn;
    }
    (*Last)->next = *First;
}

void InsertLast(PPNODE First, PPNODE Last, int No)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));
    newn->next = NULL;
    newn->data = No;

    if(*First == NULL || *Last == NULL)
    {
        *First = newn;
        *Last = newn;
    }
    else
    {
        (*Last)->next = newn;
        *Last = newn;
    }
    (*Last)->next = *First;
}

void InsertAtPos(PPNODE First, PPNODE Last, int No, int iPos)
{
    PNODE newn = NULL;
    PNODE temp = NULL;
    int i = 0;
    int iLength = 0;

    iLength = Count(*First, *Last);

    if((iPos<1) || (iPos>iLength+1))
    {
        printf("Invalid Position\n");
        return;
    }

    if(iPos == 1)
    {
        InsertFirst(First, Last, No);
    }
    else if(iPos == iLength+1)
    {
        InsertLast(First, Last, No);
    }
    else
    {
        temp = *First;
        newn = (PNODE)malloc(sizeof(NODE));
        newn->next = NULL;
        newn->data = No;

        for(i=1; i< iPos-1; i++)
        {
            temp = temp->next;
        }
        newn->next = temp->next;
        temp->next = newn;
    }
}

void DeleteFirst(PPNODE First, PPNODE Last)
{
    if((*First == NULL)|| (*Last == NULL))
    {
        printf("LL is already empty !!");
        return;
    }
    else if(*First == *Last)
    {
        free(*First);
        *First = NULL;
        *Last = NULL;
    }
    else
    {
        *First = (*First)->next;
        free((*Last)->next);
        (*Last)->next = *First;
    }
}

void DeleteLast(PPNODE First, PPNODE Last)
{
    PNODE temp = NULL;

    if((*First == NULL)|| (*Last == NULL))
    {
        printf("LL is already empty !!");
        return;
    }
    else if(*First == *Last)
    {
        free(*First);
        *First = NULL;
        *Last = NULL;
    }
    else
    {
        temp = *First;
        while(temp->next->next != *First)
        {
            temp = temp->next;
        }

        *Last = temp;
        free((*Last)->next);
        (*Last)->next = *First;
    }
}

void DeleteAtPos(PPNODE First, PPNODE Last, int iPos)
{
    PNODE temp1 = NULL;
    PNODE temp2 = NULL;
    int i = 0;
    int iLength = 0;

    iLength = Count(*First, *Last);

    if((iPos<1) || (iPos>iLength))
    {
        printf("Invalid Position\n");
        return;
    }

    if(iPos == 1)
    {
        DeleteFirst(First, Last);
    }
    else if(iPos == iLength+1)
    {
        DeleteLast(First, Last);
    }
    else
    {
        temp1 = *First;

        for(i=1; i< iPos-1; i++)
        {
            temp1 = temp1->next;
        }
        temp2 = temp1->next;
        temp1->next = temp2->next;
        free(temp2);
    }
}

int main()
{
    PNODE Head = NULL;
    PNODE Tail = NULL;
    int iRet = 0;

    InsertFirst(&Head, &Tail, 51);
    InsertFirst(&Head, &Tail, 21);
    InsertFirst(&Head, &Tail, 11);

    Display(Head, Tail);
    iRet = Count(Head, Tail);
    printf("Number of elements are : %d\n",iRet);

    InsertLast(&Head, &Tail, 111);
    InsertLast(&Head, &Tail, 121);
    InsertLast(&Head, &Tail, 151);

    Display(Head, Tail);
    iRet = Count(Head, Tail);
    printf("Number of elements are : %d\n",iRet);

    InsertAtPos(&Head, &Tail, 91, 4);

    Display(Head, Tail);
    iRet = Count(Head, Tail);
    printf("Number of elements are : %d\n",iRet);

    DeleteAtPos(&Head, &Tail, 4);

    Display(Head, Tail);
    iRet = Count(Head, Tail);
    printf("Number of elements are : %d\n",iRet);

    DeleteFirst(&Head, &Tail);

    Display(Head, Tail);
    iRet = Count(Head, Tail);
    printf("Number of elements are : %d\n",iRet);

    DeleteLast(&Head, &Tail);

    Display(Head, Tail);
    iRet = Count(Head, Tail);
    printf("Number of elements are : %d\n",iRet);

    return 0;
}
